### PR TITLE
HPCC-21700 Created optimized pow expr for ints

### PR DESCRIPTION
**Pull request by jpmcmu**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [X] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [X] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the CONTRIBUTORS document.
- [X] The change has been fully tested:
  - [X] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [X] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: ecl/hqlcpp/hqlcpp.cpp [hunk 1]
```diff
@@ -7487,6 +7487,83 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
     }
 }
 
+void HqlCppTranslator::doBuildExprPow(BuildCtx & ctx, IHqlExpression * expr, CHqlBoundExpr & tgt)
+{
+    assertex(expr->numChildren() == 2);
+    
+    IHqlExpression * left = expr->queryChild(0);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7494]: minor: left is not needed  until later, so may as well delay that until later.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 2]
```diff
@@ -7487,6 +7487,83 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+    IValue * value = right->queryValue();
+    bool notConstant = (value == nullptr);
+    if (notConstant) 
+    {
+        doBuildExprSysFunc(ctx, expr, tgt, powerId);
+        return;
+    }
+
+    const uint MAX_EXP_INLINE_VALUE = 3;
+    int64_t rightIntVal = value->getIntValue();
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7513]: trivial: We tend to use __int64 rather than int64_t.

**Comment by jpmcmu** [comment]
@ghalliday missing a change, about to push it up. Did have a few questions about the best way to handle div / cast.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 3]
```diff
@@ -7487,6 +7487,83 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+    if (abs(rightIntVal) > MAX_EXP_INLINE_VALUE)
+    {
+        doBuildExprSysFunc(ctx, expr, tgt, powerId);
+        return;
+    }
+
+    IHqlExpression* outExpr = nullptr;
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7520]: A couple of issues:
a. Expression link counts.  The case right==0 will create an expression with a link count of 1, the case right=1 is no linking the result.  The expression is not released - so it will leak for right==0 (and right != 1)
You should check changes carefully for memory leaks (can use valgrind, but need to use the --leakcheck option to ensure everything is cleaned up correctly.

b. Different kinds of expressions.  The same structures are used to represent ECL expressions and c++ expressions, but you need to be very careful about not mixing them.  So buildExpr() expects an ECL expression, but on line 7563 you are passing it a c++ expression.  That could cause problems.

I think in this case you should not be calling buildExpr at all, but should be setting tgt.expr.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 4]
```diff
@@ -7487,6 +7487,83 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+        // Is there a better way to get a temp real here?
+        //---------------------------------------------------------------------------
+
+        CHqlBoundExpr leftBound;
+        buildTempExpr(ctx, left, leftBound);
+
+        CHqlBoundExpr tempReal;
+        OwnedITypeInfo realType = makeRealType(DEFAULT_REAL_SIZE);
+        doBuildCastViaTemp(ctx, realType, leftBound, tempReal);
+        outExpr = tempReal.expr;
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7542]: Another mixture of not linking here, and linking within the loop;

File: ecl/hqlcpp/hqlcpp.cpp [hunk 5]
```diff
@@ -7487,6 +7487,83 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+
+        for (int i = 1; i < abs(rightIntVal); i++)
+        {
+            outExpr = createValue(no_mul, realType, LINK(outExpr), LINK(tempReal.expr));
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7546]: realType should be linked - it would crash if the expression wasn't leaked.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 6]
```diff
@@ -7487,6 +7487,83 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+
+        if (rightIntVal < 0)
+        {
+            CHqlBoundExpr bound;
+            buildTempExpr(ctx, outExpr, bound);
+
+            //---------------------------------------------------------------------------
+            // Question:
+            // This inserts a check for divide by zero. What is the best way to not have
+            // that check generated?
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7557]: The value could be zero, so it needs to have that check.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 7]
```diff
@@ -7487,6 +7487,83 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+
+        for (int i = 1; i < abs(rightIntVal); i++)
+        {
+            outExpr = createValue(no_mul, realType, LINK(outExpr), LINK(tempReal.expr));
+        }
+
+        if (rightIntVal < 0)
+        {
+            CHqlBoundExpr bound;
+            buildTempExpr(ctx, outExpr, bound);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7552]: again not needed - it is already a c++ expression.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 8]
```diff
@@ -7487,6 +7487,83 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+
+            //---------------------------------------------------------------------------
+            // Question:
+            // This inserts a check for divide by zero. What is the best way to not have
+            // that check generated?
+            //---------------------------------------------------------------------------
+            outExpr = createValue(no_div, realType, createConstant(1.0), bound.expr);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7559]: tgt.expr.setown(createValue(no_div, LINK(realType), createConstant(1.0), LINK(tgt.expr));

**Review by ghalliday — CHANGES_REQUESTED** [review]
@jpmcmu a few comments, main issues are making sure link counting is correct, and difference between translated(c++), and untranslated expressions (ECL)

File: ecl/hqlcpp/hqlcpp.cpp (hunk 6 above)
Comment (jpmcmu) [reply ecl/hqlcpp/hqlcpp.cpp]: Right, I was looking to move that check above the entire expansion. I ended up doing this during the code review changes, not sure if the approach I took was the best route though.

**Comment by jpmcmu** [comment]
@ghalliday implemented code review changes, please review

**Comment by ghalliday** [comment]
@jpmcmu I'll take a look.  For future reference you can click refresh on the review to re-request a review.  (I have done it).

File: ecl/hqlcpp/hqlcpp.cpp [hunk 9]
```diff
@@ -7487,6 +7487,74 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+    if (notConstant) 
+    {
+        doBuildExprSysFunc(ctx, expr, tgt, powerId);
+        return;
+    }
+
+    __int64 rightIntVal = value->getIntValue();
+    __int64 absExp = abs(rightIntVal);
+
+    const uint MAX_EXP_INLINE_VALUE = 3;
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7514]: should be unsigned not uint (it causes the windows build to fail)

File: ecl/hqlcpp/hqlcpp.cpp [hunk 10]
```diff
@@ -7487,6 +7487,74 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+    {
+        tgt.expr.setown(createConstant(left->queryType()->castFrom(1.0)));
+    }
+    else if (rightIntVal == 1)
+    {
+        tgt.expr.setown(LINK(leftBound.expr));
+    }
+    else
+    {
+        LINK(leftBound.expr);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7538]: This should be combined with the outExpr definition:

        IHqlExpression* outExpr = LINK(leftBound.expr);

But better is to use an OwnedHqlExpr so that if an exception is thrown it will be cleanly released.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 11]
```diff
@@ -7487,6 +7487,74 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+        }
+
+        tgt.expr.setown(outExpr);
+
+        if (rightIntVal < 0)
+        {
+            tgt.expr.setown(createValue(no_div, LINK(realType), createConstant(1.0), LINK(tgt.expr)));
+        }
+
+        OwnedHqlExpr cond = createBoolExpr(no_ne, LINK(leftBound.expr), getZero());
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7553]: The types of the left and right hand sizes of this expression do not match - that may trigger internal checks.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 12]
```diff
@@ -7487,6 +7487,74 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+        tgt.expr.setown(createValue(no_if, leftBound.expr->getType(), LINK(cond), LINK(tgt.expr), getZero()));
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7554]: The behaviour of division by zero can be configured. (See doBuildAssignDivide).

Also, the types of the the left and right sides of the if() do not match - real and integer.  Better would be to reuse the divide code.

File: ecl/regress/power3.ecl [hunk 13]
```diff
@@ -0,0 +1,24 @@
+person := dataset('person', { unsigned8 person_id, string1 per_sex, unsigned per_ssn, string40 per_first_name, data9 per_cid, unsigned8 xpos }, thor);
+
+output(person,
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply ecl/regress/power3.ecl:6]: Still need tests in testing/regress/ecl

File: ecl/hqlcpp/hqlcpp.cpp [hunk 14]
```diff
@@ -7487,6 +7487,74 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+    OwnedHqlExpr castLeft = ensureExprType(left, realType);
+    buildTempExpr(ctx, castLeft, leftBound);
+
+    if (rightIntVal == 0)
+    {
+        tgt.expr.setown(createConstant(left->queryType()->castFrom(1.0)));
+    }
+    else if (rightIntVal == 1)
+    {
+        tgt.expr.setown(LINK(leftBound.expr));
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7534]: tgt.set(leftBound)

is simpler

Comment (ghalliday) [reply ecl/hqlcpp/hqlcpp.cpp:7534]: Even better to avoid the special case since the general case will generate the same.

File: ecl/hqlcpp/hqlcpp.cpp (hunk 12 above)
Comment (ghalliday) [reply ecl/hqlcpp/hqlcpp.cpp:7554]: This is only needed if rightIntVal < 0.

**Review by ghalliday — COMMENTED** [review]
@jpmcmu getting closer.   A few more comments.

**Review by AttilaVamos — CHANGES_REQUESTED** [review]

File: testing/regress/ecl/power.ecl [hunk 15]
```diff
@@ -0,0 +1,12 @@
+REAL    real_const := 1.123456;
```
[the remark this answers is not in the document]
Comment (jpmcmu) [reply testing/regress/ecl/power.ecl:18]: @AttilaVamos Thanks fixed.

**Comment by jpmcmu** [comment]
@ghalliday Should I be rounding values coming from POWER in the tests to avoid floating point comparison issues between different cpu archs / compiler flags?

File: ecl/hqlcpp/hqlcpp.cpp [hunk 16]
```diff
@@ -7487,6 +7487,81 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
     }
 }
 
+void HqlCppTranslator::doBuildExprPow(BuildCtx & ctx, IHqlExpression * expr, CHqlBoundExpr & tgt)
+{
+    assertex(expr->numChildren() == 2);
+    
+    IHqlExpression * right = expr->queryChild(1);
+
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7495]: whitespace: wouldn't normally have a blank line here

File: ecl/hqlcpp/hqlcpp.cpp [hunk 17]
```diff
@@ -7487,6 +7487,81 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+        doBuildExprSysFunc(ctx, expr, tgt, powerId);
+        return;
+    }
+    
+    OwnedITypeInfo realType = makeRealType(DEFAULT_REAL_SIZE);
+    CHqlBoundExpr leftBound;
+    
+    IHqlExpression * left = expr->queryChild(0);
+    OwnedHqlExpr castLeft = ensureExprType(left, realType);
+    buildTempExpr(ctx, castLeft, leftBound);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7526]: trivial: This code can move inside the else branch.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 18]
```diff
@@ -7487,6 +7487,81 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+        for (int i = 1; i < absExp; i++)
+        {
+            outExpr.setown(createValue(no_mul, LINK(realType), outExpr.getClear(), LINK(leftBound.expr)));
+        }
+
+        tgt.expr.setown(outExpr.getClear());
+
+        if (rightIntVal < 0)
+        {
+            OwnedHqlExpr numerator = createConstant(tgt.expr->queryType()->castFrom(1.0));
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7544]: minor: use realType instead of tgt.expr->queryType()

File: ecl/hqlcpp/hqlcpp.cpp [hunk 19]
```diff
@@ -7487,6 +7487,81 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+            IHqlExpression* divisor = tgt.expr.get();
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7545]: minor: "= tgt.expr" would be more normal.  We don't normally call get unless required (e.g. in a printf)

File: ecl/hqlcpp/hqlcpp.cpp [hunk 20]
```diff
@@ -7487,6 +7487,81 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+            OwnedHqlExpr pureExpr = createValue(no_div, left->getType(), numerator.getClear(), LINK(divisor));
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7546]: bug: left->getType() is theoretically wrong - this should be realType (not sure if it currently can be non-real).

File: ecl/hqlcpp/hqlcpp.cpp [hunk 21]
```diff
@@ -7487,6 +7487,81 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+
+            CHqlBoundTarget tempTarget;
+            createTempFor(ctx, expr, tempTarget);
+
+            IValue * zero = left->queryType()->castFrom(false, 0);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7551]: bug: left->queryType() is not correct.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 22]
```diff
@@ -7487,6 +7487,81 @@ void HqlCppTranslator::doBuildAssignExecuteWhen(BuildCtx & ctx, const CHqlBoundT
...
+            OwnedHqlExpr eZero = createConstant(zero);
+
+            BuildCtx subctx(ctx);
+            IHqlStmt * stmt = subctx.addFilter(leftBound.expr);
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7555]: bug: This test needs to be on pureExpr, (but avoid re-evaluating it).  If left is 1E-1000 and right is -2, then left*left will be 0 (due to real underflow), but this test will not be triggered.

File: ecl/hqlcpp/hqlcpp.cpp [hunk 23]
```diff
@@ -7522,8 +7597,6 @@ void HqlCppTranslator::doBuildExprDivide(BuildCtx & ctx, IHqlExpression * expr,
     }
 }
 
-
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7525]: whitespace: unrelated change

File: ecl/hqlcpp/hqlcpp.cpp (hunk 17 above)
Comment (ghalliday) [inline ecl/hqlcpp/hqlcpp.cpp:7526]: minor: better to call buildSimpleExpr() than buildTempExpr().  It will reduce the number of occasions that a temporary variable is actually created.

File: testing/regress/ecl/power.ecl [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+
+OUTPUT(POWER(rVal,0), named('pow_0'));
+OUTPUT(POWER(rVal,1), named('pow_1'));
+OUTPUT(POWER(rVal,2), named('pow_2'));
+OUTPUT(POWER(rVal,3), named('pow_3'));
+OUTPUT(POWER(rVal,4), named('pow_4'));
+OUTPUT(POWER(rVal,-1), named('pow_n1'));
+OUTPUT(POWER(rVal,-2), named('pow_n2'));
+OUTPUT(POWER(rVal,-3), named('pow_n3'));
+OUTPUT(POWER(rVal,-4), named('pow_n4'));
```
Comment (ghalliday) [inline testing/regress/ecl/power.ecl:29]: This should include test cases for the unusual conditions.  POWER(0, -1), POWER(1E-1000,-2)

**Review by ghalliday — COMMENTED** [review]
@jpmcmu see comments

File: ecl/hqlcpp/hqlcpp.cpp (hunk 22 above)
Comment (jpmcmu) [reply ecl/hqlcpp/hqlcpp.cpp:7555]: Good point. Is there a way to do an epsilon comparison with addFilter?

Comment (ghalliday) [reply ecl/hqlcpp/hqlcpp.cpp:7555]: I don't think that is the way to do it - it will end up depending on which power you raise the number to, and possibly the length.  You need to be checking that the multiplied value is 0.

**Comment by jpmcmu** [comment]
@ghalliday implemented code review changes, squashed and targeted 8.6.X

File: testing/regress/ecl/key/power.xml [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+ <Row><pow_n2>0.7922967309104065</pow_n2></Row>
+</Dataset>
+<Dataset name='pow_n3'>
+ <Row><pow_n3>0.7052316520721831</pow_n3></Row>
+</Dataset>
+<Dataset name='pow_n4'>
+ <Row><pow_n4>0.6277341098113171</pow_n4></Row>
+</Dataset>
+<Dataset name='pow_5'>
+ <Row><pow_5>0</pow_5></Row>
```
Comment (AttilaVamos) [inline testing/regress/ecl/key/power.xml:29]: The generated result is 0.0

File: testing/regress/ecl/key/power.xml [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+</Dataset>
+<Dataset name='pow_6'>
+ <Row><pow_6>0</pow_6></Row>
```
Comment (AttilaVamos) [inline testing/regress/ecl/key/power.xml:32]: The generated result is 0.0

**Review by AttilaVamos — CHANGES_REQUESTED** [review]
Some mismatches between the expected (key file) and generated result.

**Comment by AttilaVamos** [comment]
@jpmcmu The "280. genjoin2, 281. genjoin3, 416. joinattr2, 425. key" errors on Thor are not related to this PR.

File: testing/regress/ecl/power.ecl [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+
+REAL    zero := 0.0 : STORED('zero');
+OUTPUT(POWER(zero, -1), named('pow_5'));
+
+REAL    nearZero := 1.0E-1000 : STORED('nearZero');
```
Comment (ghalliday) [inline testing/regress/ecl/power.ecl:34]: My mistake- this is actually unrepresentable (mixing up base 2 and base 10).  Should be 1.0E-300.

**Review by ghalliday — COMMENTED** [review]
Looks close - need to tidy up key files, add a test case, and check through comments for any missed items (e.g., unrelated newline change).

Note, this does not match code like
x := POWER(value, 2.0).
But there are relatively few instances, so I would prefer to merge as is.

**Comment by ghalliday** [comment]
> @jpmcmu The "280. genjoin2, 281. genjoin3, 416. joinattr2, 425. key" errors on Thor are not related to this PR.

@AttilaVamos I had a look at the zap report - and it looks like the weird issues are problem with the regression suite.  The zap report for one of the workunits contained results.  Is there an open issue for it?

File: testing/regress/ecl/key/power.xml [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+<Dataset name='pow_6_2'>
+ <Row><pow_6_2>0.0</pow_6_2></Row>
+</Dataset>
+<Dataset name='pow_7'>
+ <Row><pow_7>0.5</pow_7></Row>
+</Dataset>
+<Dataset name='pow_8'>
+ <Row><pow_8>8.0</pow_8></Row>
+</Dataset>
+
```
Comment (AttilaVamos) [inline testing/regress/ecl/key/power.xml:43]: This extra empty line caused the errors. Please remove it.

**Review by AttilaVamos — CHANGES_REQUESTED** [review]

**Comment by jpmcmu** [comment]
@AttilaVamos @ghalliday Key issues fixed please review

**Review by AttilaVamos — APPROVED** [review]
Looks good (in testing POW).

**Review by ghalliday — APPROVED** [review]
